assets/js/interactivity/store.ts

```diff
@@ -0,0 +1,281 @@
+ * @typedef StoreProps Properties object passed to `store`.
+ * @property {Object} state State to be added to the global store. All the
+ *                          properties included here become reactive.
+ */
+
+/**
+ * @typedef StoreOptions Options object.
+ */
+
+/**
+ * Extends the Interactivity API global store with the passed properties.
+ *
+ * These props typically consist of `state`, which is reactive, and other
+ * properties like `selectors`, `actions`, `effects`, etc. which can store
+ * callbacks and derived state. These props can then be referenced by any
+ * directive to make the HTML interactive.
+ *
+ * @example
+ * ```js
+ *  store({
+ *    state: {
+ *      counter: { value: 0 },
+ *    },
+ *    actions: {
+ *      counter: {
+ *        increment: ({ state }) => {
+ *          state.counter.value += 1;
+ *        },
+ *      },
+ *    },
+ *  });
+ * ```
+ *
+ * The code from the example above allows blocks to subscribe and interact with
+ * the store by using directives in the HTML, e.g.:
+ *
+ * ```html
+ * <div data-wp-interactive>
+ *   <button
+ *     data-wp-text="state.counter.value"
+ *     data-wp-on--click="actions.counter.increment"
+ *   >
+ *     0
+ *   </button>
+ * </div>
+ * ```
+ *
+ * @param {StoreProps}   properties Properties to be added to the global store.
+ * @param {StoreOptions} [options]  Options passed to the `store` call.
+ */
+
+type DeepPartial< T > = T extends object
+	? { [ P in keyof T ]?: DeepPartial< T[ P ] > }
+	: T;
+
+interface StoreOptions {
+	lock?: boolean | string;
+}
+
+const universalUnlock =
+	'I know using a private store means my plugin will inevitably break on the next store release.';
+
+export function store< S extends object = {} >(
+	namespace: string,
+	storePart?: DeepPartial< S >,
+	options?: StoreOptions
+): S;
+export function store< T extends object >(
+	namespace: string,
+	storePart?: T,
+	options?: StoreOptions
+): T;
+
+export function store(
+	namespace: string,
+	{ state = {}, ...block }: any = {},
+	{ lock = false }: StoreOptions = {}
+) {
+	if ( ! stores.has( namespace ) ) {
+		// Lock the store if the passed lock is different from the universal
+		// unlock. Once the lock is set (either false, true, or a given string),
+		// it cannot change.
+		if ( lock !== universalUnlock ) {
+			storeLocks.set( namespace, lock );
+		}
+		stores.set(
+			namespace,
+			new Proxy( { state: deepSignal( state ), ...block }, handlers )
+		);
+		proxyToNs.set( stores.get( namespace ), namespace );
+	} else {
+		// Lock the store if it wasn't locked yet and the passed lock is
+		// different from the universal unlock. If no lock is given, the store
+		// will be public and won't accept any lock from now on.
+		if ( ! storeLocks.has( namespace ) && lock !== universalUnlock ) {
+			storeLocks.set( namespace, lock );
+		} else {
+			const storeLock = storeLocks.get( namespace );
+			const isLockValid =
+				lock === universalUnlock ||
+				( lock !== true && lock === storeLock );
```

Comment:
Not: Maybe we could store `false` as the lock for public stores to simplify this logic a bit. But it's not important at this moment.

Reply:
Hey, that's how it already works! 😄

https://github.com/woocommerce/woocommerce-blocks/blob/ebe6f622e53f06abff104e0bf894e840e49bbe69/assets/js/interactivity/store.ts#L233

The thing is that stores locked with `lock: true` can only be unlocked with the `universalLock` (i.e., passing `true` should not unlock it!). That's why that case is covered.

Reply:
Oh, ok. Thanks 🙂